### ISSUE TITLE
Terminal shows corrupt texture sometimes when resuming the OS from a sleep state

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#69665]
**Workaround:** Press ctrl++ and ctrl+- (mac: cmd++, cmd+-) to zoom VS Code in and out which will do a complete redraw of the terminal.

---

Repro:

1. Open a terminal, run some commands
2. Sleep system
3. Resume the system (I think the delay between 2 and 3 has to be a while)

What happens exactly differs across machines, I've seen full black, full transparent and things like this:

<img width="228" alt="screen shot 2019-03-01 at 10 54 08 am" src="https://user-images.githubusercontent.com/2193314/53659411-5b467c00-3c10-11e9-9185-a63b426ce26f.png">

It seems like both the main terminal canvas texture as well as the texture atlas textures can become corrupted.

**Comment by jchv** [comment]
Does this particular flavor of issue happen outside of Linux with NVIDIA Proprietary drivers? It does seem some people report issues with terminal resume on macOS, but NVIDIA's proprietary drivers on Linux seem to have an endemic issue with sleep/resume.

**Comment by Tyriar** [comment]
I think I've seen it on my Macbook (Intel UHD Graphics 630 1536 MB) and my home PC on Windows (GTX 600 series I think).

**Comment by chrsmrtn-** [comment]
I have this too and I'm using the on board video card. Intel HD Graphics 630

**Comment by chrsmrtn-** [comment]
I resumed from a sleep state and I didn't immediately reproduce this issue.

**Comment by chrsmrtn-** [comment]
@Tyriar you mentioned a "delay between 2 and 3 has to be a while". Is that 2 to 3 hours? My sleep state was just under 2 hours

**Comment by Tyriar** [comment]
@chrsmrtn- yeah it's pretty hard to reproduce intentionally

**Comment by chrsmrtn-** [comment]
I've also noted deuteriation of the window. As in initially all is well after my computer wakes back up, but over time artifacts started appearing (using npm packages to host, monitor changes, and auto compile on change). After a while the window flashes and goes black.

**Comment by chrsmrtn-** [comment]
updated to 1809 and again ran into the issue where the terminal text was no longer visible. I could see a cursor, but interacting (even hitting enter to get the command prompt to print out) didn't work.

I closed and re-opened the terminal, but no text was displayed. I could get the command prompt to display if I hit enter

**Comment by chrsmrtn-** [comment]
note the above occurred with a working terminal and sleep was not involved. Just like my original ticket that was logged.

**Comment by achary** [comment]
same with Laptio with NVidia graphics card, UBuntu 16.04 LTS. Reloading window action is required.

**Comment by dan-dr** [comment]
@Tyriar Do you have an idea how to fix it? I could PR..

**Comment by Tyriar** [comment]
I've pushed https://github.com/Microsoft/vscode/commit/e5316826f86745cf9261dae723313a641926ac72 which might be a fix for this problem. For those seeing this a lot could you add this to your settings file and let me know if you still observe the problem:

```
"terminal.integrated.experimentalRefreshOnResume": true
```

I can make this the default behavior if it works and doesn't cause any other issues.

**Comment by Tyriar** [comment]
Has anyone had success with the `terminal.integrated.experimentalRefreshOnResume` setting?

**Comment by sagebind** [comment]
Note that like @chrsmrtn-, I see this happen randomly during normal operation as well, not just during sleep/wake.

I _do_ have an NVIDIA card, if that is the culprit. For me, I can consistently reproduce. Sleeping for any amount of time (e.g. shut laptop lid, wait 10 seconds, and open again) causes this 100% of the time for me.

I can try `terminal.integrated.experimentalRefreshOnResume` and report back.

**Comment by sagebind** [comment]
Interestingly, if I disable my NVIDIA GPU (I have an optimus laptop), the problem goes away entirely.

**Comment by Tyriar** [comment]
I'm still after some verification as to whether  `terminal.integrated.experimentalRefreshOnResume` fixes the issue, thanks.

**Comment by sagebind** [comment]
> I'm still after some verification as to whether `terminal.integrated.experimentalRefreshOnResume` fixes the issue, thanks.

I'll do my best to try it tonight, but if the name of the setting tells me anything, it may only fix the issue when it occurs on system resume, and not when it occurs while the system is running.

**Comment by Tyriar** [comment]
@sagebind can't you only repro after shutting your laptop?

> For me, I can consistently reproduce. Sleeping for any amount of time (e.g. shut laptop lid, wait 10 seconds, and open again) causes this 100% of the time for me.

I haven't heard any reports about it just happening randoml while the system is running.

**Comment by sagebind** [comment]
@Tyriar I simply mentioned the _easiest_ way for me to reproduce was via the suspend/resume method; I wasn't saying that was the _only_ way in which it happens. I added more details in the duplicate issue I opened earlier that was closed: https://github.com/Microsoft/vscode/issues/72657. To summarize, this bug triggers for me in two ways:

1. Suspend the machine with the VSCode terminal open, then resume. Triggers this behavior 100% of the time.
2. Leave the terminal open for a while with a long-running process. Not sure if there is certain kinds of process output that triggers it or not, but _sometimes_ this behavior will trigger while leaving the terminal open for a while (> 30min).

It looks like the code for the experimental setting isn't exclusively dealing with suspend/resume, so I'll try to see if I can reproduce it using the long-running method as well to see if it helps at all. Again, the second path is a lot harder for me to reproduce, but I'll try and narrow down some of the "variables" tonight.

**Comment by sagebind** [comment]
@Tyriar FWIW, it looks like @chrsmrtn- [reported it](https://github.com/Microsoft/vscode/issues/69665#issuecomment-474113053) happening for him randomly as well without suspend/resume involved:

> note the above occurred with a working terminal and sleep was not involved. Just like my original ticket that was logged.

**Comment by sagebind** [comment]
OK, I grabbed the latest Insiders build and did some testing. I've also discovered a few things that hopefully help us narrow down the problem. Firstly, I decided to test `terminal.integrated.experimentalRefreshOnResume` as @Tyriar requested to see if it helps mitigate the issue. First, testing suspend/resume behavior. Here's what I have first:

![Screenshot from 2019-04-29 21-43-04](https://user-images.githubusercontent.com/2192863/56939619-fce23080-6ace-11e9-839a-15e965263495.png)

To verify the bad behavior, I suspended the machine, and resumed immediately after suspension. Here's what I get:

![Screenshot from 2019-04-29 21-46-38](https://user-images.githubusercontent.com/2192863/56939651-27cc8480-6acf-11e9-9d06-adb5c18a9306.png)

OK, I reproduced the bad behavior (did it several times to be sure; the texture corruption varies, sometimes simply blank). Now let's try with the setting set to `true`, reload the window, and run through the suspend/resume process again:

![Screenshot from 2019-04-29 21-47-43](https://user-images.githubusercontent.com/2192863/56939687-6104f480-6acf-11e9-8395-18e15fc6978a.png)

It would appear that `terminal.integrated.experimentalRefreshOnResume` **successfully mitigates the problem on resume** by refreshing the textures pre-emptively.

The second thing I decided to do is attempt to reproduce the problem without the involvement of suspend/resume. I reset the value back to `false` in order to reproduce. This time, on a whim, I decided to follow system logs via `journalctl` to see if anything interesting showed up there. I lucked out on my intuition, because after about 20 minutes, the issue manifested itself again by itself and some graphics errors did indeed get written to the system log!

```
Apr 29 22:16:51 skie code-insiders.desktop[32735]: [17520:0429/221651.227530:ERROR:gles2_cmd_decoder.cc(18288)] [.DisplayCompositor-0x31caa6f1bc00]GL ERROR :GL_INVALID_OPERATION : glCreateAndConsumeTextureCHROMIUM: invalid mailbox name
Apr 29 22:16:51 skie code-insiders.desktop[32735]: [17520:0429/221651.229911:ERROR:gles2_cmd_decoder.cc(10129)] [.DisplayCompositor-0x31caa6f1bc00]RENDER WARNING: texture bound to texture unit 0 is not renderable. It maybe non-power-of-2 and have incompatible texture filtering.
```

And here's a screenshot of Code at this event (running two `watch` commands that I usually invoke during development when this happens):

![Screenshot from 2019-04-29 22-23-12](https://user-images.githubusercontent.com/2192863/56939917-e046f800-6ad0-11e9-9915-7147532e4023.png)

Again note that this time, **no suspend/resume** was involved. All I did was leave Code running with the terminal open, untouched, for around 20 minutes.

It's not a whole lot to go off of, but it is a start. It could be a bug in Electron perhaps, or maybe a "quirk" in the NVIDIA driver that Electron does not handle well. I am indeed using the proprietary driver for NVIDIA:

![Screenshot from 2019-04-29 22-49-50](https://user-images.githubusercontent.com/2192863/56939968-31ef8280-6ad1-11e9-86bb-2e12b96239a9.png)

**Comment by sagebind** [comment]
After further digging, this behavior sounds a lot like this Chromium bug that was never fixed: https://bugs.chromium.org/p/chromium/issues/detail?id=481398#c3. It also seems to be the same root cause as #61624. It seems like Chromium/Electron is mis-managing the lifetimes of its textures and renderers which interacts with suspend/resume in some way, which based on the little bit of GPU programming I've done would certainly result in the kind of texture glitching behavior we see here.

I've still yet to see the issue trigger again using the suspend-less approach in order to see if `experimentalRefreshOnResume` fixes the problem then as well; I figure I can do this by simply watching the system log for an error and ensuring that the error does not result in any glitches in the terminal.

**Comment by Tyriar** [comment]
@sagebind wow, great work!

So the fix was actually specific to sleep/resume, the commit has forceRedraw in paste but that was actually an accident (4889af53eae3814580425000c28b0a9963bcec1a). I think it's fine to just force the texture to recreate itself whenever you resume the OS (and get rid of the setting).

For your other problem though of this also happening randomly while running that's a bit harder to workaround, something we could do it regenerate the texture when the terminal is focused but that's probably a little too aggressive as it would cause the CPU to do a lot more work. Could you check the following when you're in the broken state in https://github.com/Microsoft/vscode/issues/69665#issuecomment-487822673?

- Check the developer tools console if there are any related errors (if so we could catch and force a redraw there)
- Set `"terminal.integrated.experimentalRefreshOnResume": true` then sleep and resume and see if everything is fixed (to verify that the same workaround is still good)

**Comment by nmoinvaz** [comment]
I do experience a similar issue where the terminal will go blank randomly on Linux. It doesn't have anything to do with suspend/resume. I am using Ubuntu 18.04.2 in a VM. It is very similar to the description in this comment https://github.com/microsoft/vscode/issues/69665#issuecomment-487814197.

**Comment by CoryM** [comment]
I have found a work around that works for me.  Setting Terminal>Integrated>Renderer Type from Auto to DOM to avoid the GPU acceleration and therefor the texture glitch in chrome.

I am also on a Linux system but team red (amd) on the GPU side.

**Comment by sagebind** [comment]
I've been busy but I intend to revisit this problem to help get it solved if I can. @Tyriar:

> - Set `"terminal.integrated.experimentalRefreshOnResume": true` then sleep and resume and see if everything is fixed (to verify that the same workaround is still good)

I didn't note this, but pretty sure I tried this and this worked. That is, after entering the broken state, sleeping and resuming to invoke the texture regeneration fixes the broken state.

> - Check the developer tools console if there are any related errors (if so we could catch and force a redraw there)

I will reproduce again with the devtools open and see what I can find. If indeed something noticeable shows up in the console then we could just regenerate the texture whenever such an error text would be logged.

**Comment by benjie** [comment]
I'm having this issue under Ubuntu 18.04 LTS using the NVIDIA proprietary drivers (NVidia 2070 Super) under KDE. This is a freshly built computer with a completely clean OS install.

The `experimentalRefreshOnResume` solution _does not_ work for me.

The zoom solution _does_ work for me.

Switching to the DOM renderer also works.

I can reliably reproduce this every single time, even from just 5 seconds of sleep; I've taken screenshots including the developer tools in case it helps. (There are no errors reported.)

After opening a clean VSCode (no other VSCode instances were open):

![Screenshot_20190721_103342](https://user-images.githubusercontent.com/129910/61589579-f4683600-aba3-11e9-9a5a-28fd96629cc8.png)

After 5 seconds of sleep, and typing a message into both the developer tools and the (invisible) terminal:

![Screenshot_20190721_103451](https://user-images.githubusercontent.com/129910/61589582-0649d900-aba4-11e9-86ec-47894e53c228.png)

Switching the renderer to DOM from canvas:

![Screenshot_20190721_103504](https://user-images.githubusercontent.com/129910/61589585-0fd34100-aba4-11e9-8d9f-b51f8823a04a.png)

Switching the renderer back to canvas still renders badly:

![Screenshot_20190721_103514](https://user-images.githubusercontent.com/129910/61589588-1bbf0300-aba4-11e9-9b91-ec5bfa7414bb.png)

But zooming up fixes it:

![Screenshot_20190721_103526](https://user-images.githubusercontent.com/129910/61589590-22e61100-aba4-11e9-9919-a602fa0630e9.png)

And zooming back down makes everything work fine again:

![Screenshot_20190721_104403](https://user-images.githubusercontent.com/129910/61589617-7f493080-aba4-11e9-8209-e60b240dc221.png)


Not sure what's going on with the increased indentation of the "Before sleep" message in the terminal :man_shrugging:

**Comment by dbaileyut** [comment]
Is there a workaround fix for this in the works on the VS Code end since the Chromium bug @sagebind found seems to have been closed without a resolution? Or a new bug filed upstream?

I run into this all the time switching between RDP and console access.

**Comment by benjie** [comment]
The zoom up/down fix seems to work, as does the DOM renderer.

**Comment by Tyriar** [comment]
Workarounds:

- Disable canvas in terminal: `"terminal.integrated.rendererType": "dom"`)
- Disable GPU acceleration: `code --disable-gpu`

**Comment by k00k** [comment]
I also experience this behavior running ElementaryOS Juno linux distro on my PC with an Nvidia 1080ti.

**Comment by sleighsoft** [comment]
When updating Nvidia drivers on Windows while a terminal is open in VSCode the terminal will be corrupted and no longer shows text. Every other view of VSCode refreshes fine after a driver update. Terminals opened after the update finishes will not be affected.

I know this is not 100% related but I believe it does not merit opening a separate issue.

**Comment by jamie-pate** [comment]
I've set "terminal.integrated.experimentalRefreshOnResume": true and this still occurs after suspend.

```
Distributor ID: Ubuntu
Description:    Pop!_OS 19.04
Release:        19.04
Codename:       disco
```
```
$ modinfo nvidia
filename:       /lib/modules/5.0.0-25-generic/updates/dkms/nvidia.ko
alias:          char-major-195-*
version:        430.34
```

```
5.0.0-25-generic
```

```Version: 1.37.1
Commit: f06011ac164ae4dc8e753a3fe7f9549844d15e35
Date: 2019-08-15T16:17:25.463Z
Electron: 4.2.7
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.31-electron.0
OS: Linux x64 5.0.0-25-generic```

**Comment by ambrop72** [comment]
Assuming the terminal is drawn OpenGL (which I am not sure), I think the proper solution would be to use the [NV_robustness_video_memory_purge](https://www.khronos.org/registry/OpenGL/extensions/NV/NV_robustness_video_memory_purge.txt) extension to be notified when resources need to be refreshed.

> PURGED_CONTEXT_RESET_NV indicates that video memory was purged
>         in a mode switch event. In that case, the application can no 
>         longer count on its resources' content being preserved, and must 
>         delete and recreate all buffer objects including FBOs and all 
>         textures bound to an FBO, transfer all relevant data again, 
>         and re-render to FBOs. Using buffer objects, including FBOs, 
>         that haven't been deleted and re-created is undefined behavior. 
>         Other objects are unaffected.

**Comment by ambrop72** [comment]
See also the Qt bug report: https://bugreports.qt.io/browse/QTBUG-56610

**Comment by Tyriar** [comment]
Another incarnation of this bug (Electron 6, Windows 10)

![image](https://user-images.githubusercontent.com/2193314/70384643-eccce200-1936-11ea-8b69-0cffe8e1ab41.png)

**Comment by Tyriar** [comment]
A related issue if you're on the WebGL renderer, you'll see a 🙁 face when the context gets lost and you cannot recover the terminal for now. This is tracked in https://github.com/xtermjs/xterm.js/issues/2253

**Comment by kevinmanify** [comment]
This happens for me as well whenever I detach/attach my second screen (v 1.42.0). Tried with 
`"terminal.integrated.experimentalRefreshOnResume": true` but still.
Zooming in and out does the trick for now.

**Comment by jeremyong** [comment]
Running into this on ArchLinux using mainline Nvidia drivers

**Comment by basilky** [comment]
`"terminal.integrated.experimentalRefreshOnResume": true `works for active terminal only. Other terminal tabs are still affected.

**Comment by memtech3** [comment]
running into this on ubuntu 20.04 nvidia proprietary drivers

**Comment by fahad-israr** [comment]
was working fine on Ubuntu 18.04 but now has "corrupt texture issue after waking from sleep"  after upgrading to Ubuntu 20.04.

**Comment by sam-cross** [comment]
Happening on Windows 10 64 bit build Version 10.0.18363.778 [November 2019 update], when waking from hibernation.

![image](https://user-images.githubusercontent.com/6699307/81748380-ae1cf400-94a1-11ea-8b20-d32f9f6d04a3.png)

Interestingly, if nothing is running in the terminal window when the system is resumed, if you press Enter a few times when focused on the terminal, it redraws and displays all the text again. However, **this doesn't happen** when running anything in the terminal, even if you Ctrl+C to end it and repeatedly press Enter. HOWEVER, the terminal DOES seem to render characters which have previously not been displayed by it (but I haven't *confirmed* this).

Just thought I'd add my experiences in case it helps narrow this down at all.

**Comment by stephan-t** [comment]
This issue occurs 100% of the time when resuming the system from suspend with Ubuntu 20.04 and NVIDIA proprietary drivers. I've also seen similar texture corruption in Chromium browser but at a much lower frequency. So the issue in VS Code seems to stem from Electron / Chromium.

Based on the given workaround, I created a [script](https://gist.github.com/stephan-t/561d0e473dddb270be3da6a8ca994306) to automatically redraw the terminal when resuming from suspend. This has only been tested on Ubuntu 20.04 so it may require slight modifications for other versions or distros.

**Comment by codebrainz** [comment]
I've seen this in Kicad (wxWidgets/gtk) and Geany (gtk) as well as a [bug report for KDE](https://bugs.kde.org/show_bug.cgi?id=364766#c67). It seems like this is not a VSCode-specific bug but rather with NVidia proprietary driver/X11.

I found a workaround [here](https://forums.developer.nvidia.com/t/gnome-shell-crashes-and-rendering-problems-on-xserver-with-430-26/77317/16) but I didn't test it myself since I dislike messing around with graphics drivers stuff on Linux, it's too much hassle for a minor bug like this.

**Comment by sam-cross** [comment]
I'm using Intel HD Graphics - 1st gen. It's not an Nvidia issue.

**Comment by f3rm4rf3r** [comment]
> This issue occurs 100% of the time when resuming the system from suspend with Ubuntu 20.04 and NVIDIA proprietary drivers. I've also seen similar texture corruption in Chromium browser but at a much lower frequency. So the issue in VS Code seems to stem from Electron / Chromium.
> 
> Based on the given workaround, I created a [script](https://gist.github.com/stephan-t/561d0e473dddb270be3da6a8ca994306) to automatically redraw the terminal when resuming from suspend. This has only been tested on Ubuntu 20.04 so it may require slight modifications for other versions or distros.

Working for me in Linux Mint 19.3 (18.04 Ubuntu based). Thanks @stephan-t

**Comment by Fmstrat** [comment]
Hi all, here's a version of the great script from @stephan-t but without the `jq` requirement and this also works with dotfiles (symlinks) as it does not replace the `settings.json` file, it uses `sed -i` to replace in-place.

https://gist.github.com/Fmstrat/7de01996206537b9dbc8f673b2c5ced2

**Comment by isum** [comment]
Same thing for me. After sleep terminal and rich scrollbar become corrupted as well as chromium dev-tools.
Linux 5.4, Manjaro, video-nvidia-440xx
PS: Before I was using a video-linux driver and everything was fine.

**Comment by JanJamaszyk-FlyNow** [comment]
Same issue with Ubuntu 20.04 and nvidia-driver-440/focal-updates,focal-security,now 440.100-0ubuntu0.20.04.1 amd64

**Comment by lal12** [comment]
same as @JanJamaszyk-FlyNow

**Comment by brainvader** [comment]
I also ran across same issue. 

My environment:

VSCode: 1.46.1
Ubuntu: 20.04 LTS
Linux Kernel: 5.4.0-40-generic

and also uses nvidia-driver 440 (proprietary, tested) mentioned above and zsh as default shell. There are no error on the developer console. I have never seen this in macOS version.

**Comment by jepperaskdk** [comment]
Ubuntu 18.04
Version 1.47.0

![image](https://user-images.githubusercontent.com/10128260/87704107-26eb4580-c79c-11ea-92ac-3f150829d209.png)

**Comment by interlark** [comment]
Happens all the time to me. Text just disappears and background color fills black:
![image](https://user-images.githubusercontent.com/20641837/88454537-d2c21e80-ce89-11ea-80a6-a1f85f7bd35a.png)

**Comment by LorenzoCerda** [comment]
Hi.
   Exactly the same problem here. 
Ubuntu 18.04
Kernel 5.4.0-42-generic
VSCode version: 1.47.2
Also with Nvidia card (although in another computer, not so long ago, it worked ok. That notebook also has a Nvidia card). 

  Thank you in advance.

**Comment by NicolaiMogensen** [comment]
> Happens all the time to me. Text just disappears and background color fills black:
> ![image](https://user-images.githubusercontent.com/20641837/88454537-d2c21e80-ce89-11ea-80a6-a1f85f7bd35a.png)
> 
> Ctrl ++, Ctrl +- Indeed fixes the issue, thank you.

This also fixes it for me. Windows 10, VSCode version: 1.47.2. Nvidia driver 441.87 on a 2080 TI RTX over Thunderbolt eGPU

**Comment by boeckMt** [comment]
> I've pushed [e531682](https://github.com/microsoft/vscode/commit/e5316826f86745cf9261dae723313a641926ac72) which might be a fix for this problem. For those seeing this a lot could you add this to your settings file and let me know if you still observe the problem:
> 
> ```
> "terminal.integrated.experimentalRefreshOnResume": true
> ```
> 
> I can make this the default behavior if it works and doesn't cause any other issues.

I have the feeling that **"terminal.integrated.experimentalRefreshOnResume": true** helps a bit but it is not completely like it was sometime before this Issue occurred.

**Comment by lal12** [comment]
It happens to me on every resume after a hibernation/standby. So when it occured I retrieved the terminal instance via dev tools. And I called `termInstance._xterm.setOption('rendererType', 'dom');` as @Tyriar did in his commit and it works. So I guess it would just be a matter of detecting a resume and calling this function?

**Comment by lal12** [comment]
I also tried `terminalInstance._xterm.refresh(...)` (and therefore `terminalInstance.forceRedraw`) which does not help.

I directly used `"terminal.integrated.rendererType": "dom"` in the settings and the issue does not occur, it only occurs for `canvas` and `experimentalWebgl`. Though `termInstance._xterm.setOption('rendererType', 'dom');` and `termInstance._xterm.setOption('rendererType', 'canvas');` also fixes the issue.

**Comment by lal12** [comment]
I tried the same thing with the xtermjs.org demo in Chrome, Chromium and Firefox. Both Chrome and Chromium have the same issue as vscode, while firefox works. So I guess it has something to do with how the chromium web engine is handling the canas buffers after a resume.

**Comment by danilo-medeiros** [comment]
I could fix following these steps: https://medium.com/@iamstorm/vscode-bug-on-linux-7ca4c4544d24

**Comment by manav28** [comment]
The fix suggested by @danilo-medeiros doesn't work for me

**Comment by LorenzoCerda** [comment]
> The fix suggested by @danilo-medeiros doesn't work for me

Unfortunately, for me that fix doesn't work either. 
Ubuntu 18.04. 
Nvidia card.

**Comment by Heath123** [comment]
I get this everywhere in the system

**Comment by anton-matosov** [comment]
I have this issue on both macOS (terminal only) and Ubuntu 18.04 with NVidia GTX960, drivers version 440.100 (entire screen, same happens to Chrome on linux only)
Zoom in/Zoom out workaround works on both platforms for VSCode, but, surprisingly, not with Chrome.

**Comment by Tyriar** [comment]
I've believe I've fixed this for the webgl renderer with https://github.com/microsoft/vscode/pull/106203, try out in the next Insiders build. If you're not on the webgl renderer yet you can switch to it using this:

```json
"terminal.integrated.rendererType": "experimentalWebgl"
```

The plan is to remove the canvas renderer (default) eventually https://github.com/microsoft/vscode/issues/106202.

**Comment by nitzanorchestra** [comment]
still having this problem, I'm on linux arch.

**Comment by fld91** [comment]
having this issue on chrome,terminal and vscode....os -ubuntu 20.04

**Comment by deepakraina99** [comment]
I am also having this issue on vscode. 
Ubuntu 20.04 OS with Nvidia graphic card. 
Though I fixed the chrome issue by enabling Vulkan as graphics backends from chrome://flags/

**Comment by grezxune** [comment]
I have had this issue with Slack, Chrome, VSCode, and some buggy stuff with Spotify. All of these happen to be snap apps except Chrome. I'm not sure if that holds any significance, but I figured it wouldn't hurt to point out.

**Comment by asetianegara** [comment]
same problem with my ubuntu 20.04 laptop. it also affect slack desktop.

i think the problem is with chrome/chromium render as vscode and slack are electron apps.

**Comment by bullsei** [comment]
I think so too, having same problems in chrome Version 86.0.4240.75 and just noticed the glitch in Vscode after resuming. I think for me i first noticed the bug in chrome 85.

**Comment by brendandebeasi** [comment]
Same here on Ubuntu 20.04. Happens to Slack, Chrome and anything using a webview (Webkit / Chromium?) renderer. I was able to solve the problem by force quitting the app through System Monitor. Not a perfect solution, but better than restarting

**Comment by bartoszwalicki** [comment]
I have this problem with Arch Linux + GNOME + NVIDIA GPU, in all chrome-based apps. I managed to fix it only in Chrome browser (VS code still messy after wake up) by adding this to startup parameters:

```--use-gl=desktop```

maybe this help somehow?
As temporary quick fix after wake up I execute `command pallete -> Developer: Reload Window`. As a drawback - it terminates all terminals.

**Comment by UnwishingMoon** [comment]
On ubuntu 20.04 with NVIDIA GPU the problem seems to appear on all chrome-based apps. To fix it (let's say it's more a workaround) I disabled the hardware acceleration on VS Code and Chrome.

To disable in VS Code you can do this:

- Open the Command Palette (Ctrl+Shift+P)
- Write: `Preferences: Configure Runtime Arguments `
- Add `"disable-hardware-acceleration": true`
- Restart VS Code

Credits to: https://stackoverflow.com/a/58762168

**Comment by sguillia** [comment]
The terminal is not the only corrupted element. See white square on the left.

Is this another issue?

![Capture du 2020-10-27 20-37-42](https://user-images.githubusercontent.com/15984272/97353959-ff287500-1894-11eb-8186-21efbf623184.png)

**Comment by deepakraina99** [comment]
@sguillia  Yes I have the same issue

**Comment by deepakraina99** [comment]
> * e

It worked for me only once or twice, after that the same thing happens. Can you check again, please?

**Comment by UnwishingMoon** [comment]
> > On ubuntu 20.04 with NVIDIA GPU the problem seems to appear on all chrome-based apps. To fix it (let's say it's more a workaround) I disabled the hardware acceleration on VS Code and Chrome.
> > To disable in VS Code you can do this:
> > 
> > * Open the Command Palette (Ctrl+Shift+P)
> > * Write: `Preferences: Configure Runtime Arguments `
> > * Add `"disable-hardware-acceleration": true`
> 
> It worked for me only once or twice, after that the same thing happens. Can you check again, please?

To me never happened again, I tried (I have a laptop) even closing the lid right away, locking the pc and waiting for the sleep and just waiting for the sleep. Neither vscode, the terminal or the minimap was bugged after returning from sleep.

**Comment by Tyriar** [comment]
The `"disable-hardware-acceleration": true` workaround should work as this issue revolved around GPU memory getting released/corrupted on sleep without giving us a hook to know about it. Note that you will obviously lose (GPU) hardware acceleration which means the CPU will be harder at work rendering things. This may not be a problem depending on your setup but it's worth mentioning.

Can anyone confirm whether switching to the webgl renderer fixed the problem? (see https://github.com/microsoft/vscode/issues/69665#issuecomment-687984308)

**Comment by lal12** [comment]
@Tyriar to quote myself from 12 Aug:
> I directly used `"terminal.integrated.rendererType": "dom"` in the settings and the issue does not occur, it only occurs for `canvas` and `experimentalWebgl`. Though `termInstance._xterm.setOption('rendererType', 'dom'); termInstance._xterm.setOption('rendererType', 'canvas');` (switching back and forth to 'canvas' rendering) also fixes the issue.

**Comment by Tyriar** [comment]
@la12 right that's correct, this never happened with "dom" as it doesn't leverage the GPU. I just want to confirm whether this is indeed fixed when it's set to experimentalWebgl, which was some work I did on the 6th of Sept. If you could test it out that would be great as I always had difficulty reproducing this.

**Comment by lal12** [comment]
@Tyriar sorry forgot it. Just tested it with the newest insiders version (1.51.0-insiders) and in the default case (which I guess is canvas) this problem occurs, but with "experimentalWebgl" this does not occur anymore!

**Comment by Tyriar** [comment]
Great! So this will probably not end up getting fixed then as the canvas renderer is going to be removed at some point (https://github.com/microsoft/vscode/issues/106202). The recommended fix for this is to use one of the alternative renderers:

```jsonc
// Slower but feature complete
"terminal.integrated.rendererType": "dom",
// Faster but missing underline styling and context restoration
"terminal.integrated.rendererType": "experimentalWebgl"
```

**Comment by ktran0918** [comment]
I can confirm using `experimentalWebgl` fixes the artifacts in the minimap and terminal. I am using a GTX 1070 on Linux and GNOME.

**Comment by ktran0918** [comment]
@Tyriar are there differences in hardware usage between `canvas` and `experimentalWebgl` we should know about?

**Comment by Tyriar** [comment]
@ktran0918 webgl is basically the next iteration of it, using lower level programming and shaders to avoid the many round trips to the GPU that the canvas renderer requires.

Main differences:

- Uses webgl to render the terminal vs a [2d context](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D) & [`drawImage`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage).
- Webgl cannot render underlines on text current https://github.com/xtermjs/xterm.js/issues/2251, note that link underlines work, just not when the shell/program decorates text with an underline.
- Webgl doesn't handle context loss well, this happens when too many terminals are open and may happen in other cases where Chromium or the GPU frees the contexts for some other reason https://github.com/xtermjs/xterm.js/issues/2253, I've seen this happen a maybe 2 times after using it for like a year.

**Comment by sguillia** [comment]
For information, this also happens when the Linux OOM killer kills vscode.

```
[240416.680580] Out of memory: Kill process 7212 (code-insiders) score 208 or sacrifice child
[240416.680595] Killed process 7212 (code-insiders) total-vm:916892kB, anon-rss:131780kB, file-rss:21216kB, shmem-rss:4892kB
[240416.711358] oom_reaper: reaped process 7212 (code-insiders), now anon-rss:0kB, file-rss:25804kB, shmem-rss:4888kB
```

Terminal then shows corrupt texture.

**Comment by Hetty82** [comment]
![Screenshot 2020-11-18 at 17 37 35](https://user-images.githubusercontent.com/2639069/99559227-d5aac700-29c4-11eb-856d-cc5499dfab86.png)
I have this outside of the terminal

**Comment by m3l1x** [comment]
I'm also getting corrupted graphics in VSCode terminal and Discord on Ubuntu 20.10 in VMware Player 16.0.0 and 16.1.0. The interesting thing about it is, that on Ubuntu 20.04 everything is perfectly fine within same environment.

**Comment by justdanyul** [comment]
I'm seeing this all over the place, I've seen this issue in VSCode, Obsidian and Anki. Noice, transparent and black rectangles after prolonged sleep.

I'm on Nvidia closed source drivers, and I've seen Nvidia mentioned here and other places before, I was wondering, can people in this thread give a thumbs up to this message if they are on Nvidia and seeing this, or thumbs down if they are on Intel/AMD and seeing this issue?

**Comment by robbins** [comment]
I thought that [this](https://wiki.archlinux.org/index.php/NVIDIA/Tips_and_tricks#Preserve_video_memory_after_suspend) could potentially fix the issue, but it doesn't look like it works, assuming I enabled it correctly. Maybe someone else can try it if interested.

**Comment by alexandis** [comment]
I have it on Windows 10, videocard NVidia GeForce MX250

**Comment by pstewart85** [comment]
Hi all, 

been experiencing this glitching on the file explorer and even out around VScode itself.  At first it showed up only when using certain specific icon packs (Material Icon Pack), switching from them removed the glitching.  Further inspection showed the glitching is also occurring in the developer tools window that doesn't have the icons

The symptoms appear after waking form a sleep state. 

Attempted the work around, does not appear to work.  Hard restart of the machine will fix the glitching.

```
Macbook Pro 16" 
AMD Radeon Pro 5300M GPU

VSCode 1.51.1
Commit: e5a624b788d92b8d34d1392e4c4d9789406efe8f
Date: 2020-11-11T01:11:34.018Z (3 wks ago)
OS: Darwin x64 19.6.0
```

![Screen Shot 2020-12-08 at 9 43 22 AM](https://user-images.githubusercontent.com/9097676/101533241-32adf300-395b-11eb-967d-d9dfa57dfcd6.png)

![Screen Shot 2020-12-08 at 1 11 58 PM](https://user-images.githubusercontent.com/9097676/101533327-4d806780-395b-11eb-8855-e40040218d88.png)

**Comment by damyan90** [comment]
Experienced the same on Ubuntu 20.04 with nvidia-driver-450. Workaround with enabling "Vulcan" in chrome://flags did the trick, but decreased performance dramatically.

After installing yesterday the **nvidia-driver-455** it seems to be working without any workarounds. I noticed the issue right after resuming the laptop from suspend mode but then it was somehow "refreshed" and got back to normal. 
Give it a try ;)

**Comment by okonon** [comment]
I am seeing this on my mac:
![image](https://user-images.githubusercontent.com/1832398/104463487-d81e3d00-557f-11eb-8d8d-20a228f61e18.png)

![image](https://user-images.githubusercontent.com/1832398/104463387-b624ba80-557f-11eb-9fef-2f31bac50fd0.png)

**Comment by hect1c** [comment]
Having the same issue on mac. 

OS version: 11.1
VS code: 1.52.1

Disabling hardware acceleration also worked for me in the argv.json file `disabled-hardware-acceleration: true`

**Comment by Mohammad-Mohsen** [comment]
@Tyriar
You said it's long and hard to reproduce intentionally, but I can reproduce it in 3 seconds in my system (Ubuntu 20.04).
1. Open VS Code (If it is not already)
2. According to your system, go to switch users screen, relog in from there (you don't need to log out, just re-enter to your account from switch users screen)
And there you go

**Comment by kamenminkov** [comment]
I still think that disabling hardware acceleration is not a permanent solution but rather just a way to prove that the issue is something related to that.

**Comment by okonon** [comment]
I agree as well. This is a workaround

**Comment by Tyriar** [comment]
We will be changing the default to the webgl renderer in the next version (https://github.com/microsoft/vscode/issues/118064) which should see this fixed automatically for most people as it doesn't happen there. We will still keep the canvas renderer around and fallback to it from the webgl renderer if webgl2 is not supported (certain hardware, Codespace on Safari, maybe some mobile browsers, etc.)

**Comment by realvictorprm** [comment]
glad to see a solution in sight @Tyriar , many thanks for informing us all here!

I have this bug now since 2 years with VSCode and my Ubuntu work laptop :(

**Comment by realvictorprm** [comment]
Well, just installed the insiders version and did send my pc to sleep for testing and it didn't work. The WebGL Context was lost. Nvidia isn't great at creating proper working drivers for my laptop it appears ._.

**Comment by bsdz** [comment]
With the latest update (1.55.0) where "WebGL renderer is the new default" I found this terminal issue being replaced with a terminal that crashes (leaving a sad face icon) after resuming from hibernation.

I can get the old behaviour back by setting my "Renderer Type" to "experimentalWebgl" then using the "CTRL-+" & "CTRL--" workaround as described above.

Perhaps the "auto" setting for renderer type is choosing canvas or dom and the chosen option doesn't support hibernation.

**Comment by iamparasbansal** [comment]
I read all the discussions above and the issue is still existing in my Lenovo Laptop. I am using Nvidia Geforce 710 graphics and I am a Ubuntu 20.04 user. 
The VS Code screen goes half black after I wake it up from sleep. 

![VSCODE corrupt](https://user-images.githubusercontent.com/67871592/114208329-f75c1900-997a-11eb-8214-ebb215a0b9a0.png)

I can't turn off my hardware acceleration due to multiple reasons. Is there any fix available to this problem??

**Comment by Tyriar** [comment]
@bsdz in insiders we've changed to handle webgl context losses to drop down to the dom renderer, so I believe that is worked around for now.

**Comment by lal12** [comment]
I am using the new renderer for a while, the sleep/hibernation issue is occurring at least less often but still like half the time. But additionally all in all it is much less stable, sometimes getting a scrambled image just when pulling it up after a few hours from the background. Or quite often even a crashed terminal (sad smiley icon). So compared to the old renderer where I had to do zoom in+zoom out (ctrl++,ctrl+-) I am not sure if the overall experience is the better one...

**Comment by pyr-kaboosh** [comment]
Same issue on nvidia, manjaro, xfce. ctrl + + and ctrl + - fixes it, or closing the terminals and opening a new one.

**Comment by cheyngoodman** [comment]
Reproducible with LightDM via `dm-tool lock` on nvidia, manjaro, i3.

The resize hotkeys work as a workaround to not lose your sessions. Thanks @pyr-kaboosh !

A similar issue occurs with other apps in Chromium or Electron, but those apps resolve their graphical artifacts after a few seconds while Code requires this workaround.

**Comment by edupsousa** [comment]
Thanks, having trouble with this issue for years, just installed the Insiders version and the problem is gone! Looking forward to see it in production.

Linux Mint 20.01 - Cinnamon
GeForce GTX 1050 - Nvidia Driver Version 450.119.03

**Comment by interlark** [comment]
Fix with Ctrl++ and Ctrl+- doesn't work anymore.
![image](https://user-images.githubusercontent.com/20641837/118086986-44019c80-b3de-11eb-83cd-e3b15182b8f2.png)

**Comment by Tyriar** [comment]
@interlark what version are you on?

**Comment by interlark** [comment]
@Tyriar 
```
Version: 1.55.2
Commit: 3c4e3df9e89829dce27b7b5c24508306b151f30d
Date: 2021-04-13T09:37:02.931Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Linux x64 5.8.0-50-generic
```

**Comment by Tyriar** [comment]
@interlark hmm, that version is meant to handle context loss properly 🤔 https://github.com/microsoft/vscode/issues/120393

**Comment by eox2** [comment]
Version: 1.56.0
Commit: cfa2e218100323074ac1948c885448fdf4de2a7f
Date: 2021-05-04T21:58:14.757Z
Electron: 12.0.4
Chrome: 89.0.4389.114
Node.js: 14.16.0
V8: 8.9.255.24-electron.0
OS: Linux x64 5.4.0-73-generic

Am on Linux Mint Cinnamon if it helps with Nvidia 1050.

The Ctrl++ seems like it worked for me.

**Comment by jesusgn90** [comment]
still happening on:

```
Version: 1.57.0
Commit: b4c1bd0a9b03c749ea011b06c6d2676c8091a70c
Date: 2021-06-09T17:18:42.895Z
Electron: 12.0.9
Chrome: 89.0.4389.128
Node.js: 14.16.0
V8: 8.9.255.25-electron.0
OS: Linux x64 5.4.0-7642-generic
```

and the workaround `ctrl + and ctrl -` works for me but it's a workaround...

**Comment by RahulPalve** [comment]
Having same issue, 
Ubuntu 20.04
NVIDIA MX150
In my case the issue occurs when I turn on dedicated graphics card, everything was working fine Intel's onboard graphics.

**Comment by rdanter** [comment]
I too am seeing this in multiple areas of the VSCode window.

I'm using Kubuntu 20.04
NVidia GT218M driver340.108

Seems to happen pretty much every time the screen sleeps (even if the rest of the laptop does not)

**Comment by luanmm** [comment]
This happens to me after some time (not just when asleep). Don't know why exactly, but seems to be on a screen refresh (maybe triggered by graphics hardware). I'm on an Pop!_OS 21.04, with an NVIDIA GeForce MX110 graphics driver.

This is the most annoying thing that happens with VS Code in my daily use! :roll_eyes:
Hope this information could help somehow :smile:

**Comment by alioguzhan** [comment]
I can verify that this was happening to me with `Nvidia GT 1030`. Then I replaced it with a `Radeon RX 560`. I haven't seen that problem since then.

Ubuntu 20.04.2

**Comment by YigitDemirag** [comment]
Same issue persists with `RTX 1080Ti`

```


CPUs | Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz (16 x 4700)
-- | --
GPU Status | 2d_canvas: enabled gpu_compositing: enabled multiple_raster_threads: enabled_on oop_rasterization: disabled_off opengl: enabled_on rasterization: disabled_software skia_renderer: enabled_on video_decode: disabled_software vulkan: disabled_off webgl: enabled webgl2: enabled
Load (avg) | 2, 2, 2
Memory (System) | 31.21GB (18.65GB free)
Process Argv | --no-sandbox --unity-launch --crash-reporter-id c9c42855-6208-4a92-a148-e85e2e5bb3ab
Screen Reader | no
VM | 0%

```

**Comment by krkarel** [comment]
This issue persists for me as well. This happens every time the computer resumes from sleep and is only resolved by terminating the shell and opening a new one. Artifacts also appear and build up over time during normal use.
```
System:    Kernel: 5.4.0-80-generic x86_64 bits: 64 compiler: gcc v: 9.3.0 Desktop: Cinnamon 4.8.6 
           wm: muffin dm: LightDM Distro: Linux Mint 20.1 Ulyssa base: Ubuntu 20.04 focal 
Machine:   Type: Desktop System: Dell product: Precision Tower 5810 v: 01 serial: <filter> 
           Chassis: type: 7 serial: <filter> 
           Mobo: Dell model: 0K240Y v: A02 serial: <filter> UEFI [Legacy]: Dell v: A07 
           date: 04/14/2015 
CPU:       Topology: 6-Core model: Intel Xeon E5-1650 v3 bits: 64 type: MT MCP arch: Haswell 
           rev: 2 L2 cache: 15.0 MiB 
           flags: avx avx2 lm nx pae sse sse2 sse3 sse4_1 sse4_2 ssse3 vmx bogomips: 83815 
           Speed: 1198 MHz min/max: 1200/3800 MHz Core speeds (MHz): 1: 1198 2: 1198 3: 3604 
           4: 1200 5: 1197 6: 1197 7: 3719 8: 1198 9: 1197 10: 1198 11: 1203 12: 1198 
Graphics:  Device-1: NVIDIA GK208B [GeForce GT 710] vendor: eVga.com. driver: nvidia v: 460.91.03 
           bus ID: 07:00.0 chip ID: 10de:128b 
           Display: x11 server: X.Org 1.20.9 driver: nvidia 
           resolution: 1368x768~60Hz, 1680x1050~60Hz, 1366x768~60Hz 
           OpenGL: renderer: GeForce GT 710/PCIe/SSE2 v: 4.6.0 NVIDIA 460.91.03 direct render: Yes 
```

**Comment by saedx1** [comment]
I'm not an expert on this. But I have the same problem, and I think this is related to how the graphics card (nvidia in particular?) behaves when going to sleep or so (at least for me it happens when resuming from a sleep).

**Comment by cstrahan-blueshift** [comment]
> This happens to me after some time (not just when asleep). Don't know why exactly, but seems to be on a screen refresh (maybe triggered by graphics hardware). I'm on an Pop!_OS 21.04

Same here. I'm on an Oryx Pro laptop ([oryp7](https://tech-docs.system76.com/models/oryp7/README.html)) from System76. I can provide other details, if desired. 

I can currently work around via `Ctrl`+`+`/`Ctrl`+`-`.

**Comment by ElementalWarrior** [comment]
@cstrahan-blueshift I love you. Zoom reset works wonders!

**Comment by kristiankielhofner** [comment]
In case it's helpful for anyone else I was able to solve this with the following:

https://gist.github.com/bmcbm/375f14eaa17f88756b4bdbbebbcfd029

Needs some tweaks depending on driver version, distro packages vs binary install, etc but solves the issue with VS Code and various other applications. For some setups it's also worth checking out disabling nvidia-persistence:

https://wiki.archlinux.org/title/NVIDIA/Tips_and_tricks#Driver_persistence

**Comment by Tyriar** [comment]
Since a lot of people are still hitting this and we went back on our decision to remove the canvas renderer (it will always be there as a fallback), we'll upstream an API to clear that texture atlas so it will get fixed there too: https://github.com/xtermjs/xterm.js/issues/3455

**Comment by connor4312** [comment]
I've never seen this so not sure how to repro, any steps?

**Comment by meganrogge** [comment]
I think it's fine to say this is verified because we now detect if the OS is resuming from sleep and clear the texture to prevent this from happening

**Comment by MartinWallgren** [comment]
I can verify that this is solved. I used to see this everytime i locked/unlocked my computer. It's not happening anymore.

**Comment by egaznep** [comment]
The terminal I left open before suspended my PC
![image](https://user-images.githubusercontent.com/17089813/148923448-4faafcb5-1f7d-4b71-862d-3567a15b2b2f.png)

Things are good if I _start a new terminal_ using (+)
![image](https://user-images.githubusercontent.com/17089813/148923804-94ead574-d1d6-4890-9e80-2b7c1c0c7641.png)

Console log upon inspection
![image](https://user-images.githubusercontent.com/17089813/148923161-b8c99252-b0c9-4a4a-8331-383be83f8775.png)

I also have the same issue on my PC with Ubuntu 20.04.2, NVIDIA proprietary drivers but I see a different console log upon inspection of the terminal. Similar to what @sguillia experienced I can zoom in to get the contents back.

**Comment by simontims** [comment]
I see this in 1.64.2 via Remote Desktop with the PowerShell Integrated Console.

1) RDP to a Windows 10 Pro machine running VSCode
2) Have a ps1 file open
3) Have some content in the terminal window
4) Close RDP (don't log out, just close it)
5) Reconnect to the same session. Terminal window is blank with only the cursor present. Cursor responds to carriage returns, clear [return] places the cursor back to the top etc, but no text is visible
6) Ctrl +- Ctrl ++ fixes the issue and reveals text
7) Close the Integrated Console and click Yes when asked if you want to relaunch also fixes this

**Comment by cmey** [comment]
Still annoying

VSCode:
```
Version: 1.66.2
Commit: dfd34e8260c270da74b5c2d86d61aee4b6d56977
Date: 2022-04-11T07:49:24.808Z
Electron: 17.2.0
Chromium: 98.0.4758.109
Node.js: 16.13.0
V8: 9.8.177.11-electron.0
OS: Linux x64 5.16.15-76051615-generic
```

OS: Linux Pop!_OS 21.10

GPU: Nvidia 1080 w/ nvidia-driver-510

Easy to repro (100% reproducible):
 - Suspend the computer, and upon return the integrated terminal display is corrupt (looks all black for me).
 - Zooming (in or out) fixes it.

**Comment by codebrainz** [comment]
Since I upgraded to Ubuntu 22.04, my vscode terminal doesn't show the random coloured pixels anymore but text in the console (it seems apart from those fancy ohmyposh prompts) goes missing with the occasional character showing up randomly more like @egaznep first screenshot above).

Not sure if it helps, but I have also noticed since the upgrade, Firefox's client area (but not toolbar/tabs/wm frame) started doing the random colours just like the screenshot in the OP when resuming from sleep, or sometimes it just has garbage textures on it, which go away when you click in the client area.

It seems like it's maybe not just a vscode/electron/chrome issue but something with graphics drivers or window server or such.

Ubuntu 22.04, using GNOME with X11 session, NVidia GeForce GTX 970 with proprietary driver (470)

**Comment by Tyriar** [comment]
@codebrainz that was a regression that has since been fixed.

Since there haven't been any comments saying this still happens since April I'm going to close this off. Let me know if you see this, and if so run the new Show Terminal Texture Atlas command and upload the image is puts in your workspace root:

![image](https://user-images.githubusercontent.com/2193314/207088985-7605d9b6-61e5-4ef4-ac31-8d9d67ae9e50.png)

**Comment by ionics** [comment]
Hi folks, this still seems to be an issue. Zooming IN or OUT fixes it.

After suspend, terminal output broken.

![image](https://user-images.githubusercontent.com/3054517/210509569-1c0c33cf-f70a-4df9-8559-9896631003f4.png)

![image](https://user-images.githubusercontent.com/3054517/210510102-08c8bc59-4649-4024-870a-d160eb4bacf1.png)

System: Ubuntu 20.04 LTS AMD Ryzen 5 - NVIDIA GeForce GTX 1060 6GB
Linux username 5.15.0-53-generic #59~20.04.1-Ubuntu SMP Thu Oct 20 15:10:22 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

VGA compatible controller: NVIDIA Corporation GP106 [GeForce GTX 1060 6GB] (rev a1) (prog-if 00 [VGA controller])
        Subsystem: NVIDIA Corporation GP106 [GeForce GTX 1060 6GB]
        Flags: bus master, fast devsel, latency 0, IRQ 80
        Memory at f6000000 (32-bit, non-prefetchable) [size=16M]
        Memory at e0000000 (64-bit, prefetchable) [size=256M]
        Memory at f0000000 (64-bit, prefetchable) [size=32M]
        I/O ports at e000 [size=128]
        Expansion ROM at 000c0000 [virtual] [disabled] [size=128K]
        Capabilities: <access denied>
        Kernel driver in use: nvidia
        Kernel modules: nvidiafb, nouveau, nvidia_drm, nvidia
